arraycontext/context.py

```diff
@@ -304,25 +304,26 @@ def zeros_like(self, ary: Array) -> Array:
 
     @abstractmethod
     def from_numpy(self,
-                   array: Union["np.ndarray[Any, Any]", ScalarLike]
-                   ) -> Union[Array, ScalarLike]:
+                   array: ArrayOrContainerOrScalarT
+                   ) -> ArrayOrContainerOrScalarT:
         r"""
         :returns: the :class:`numpy.ndarray` *array* converted to the
             array context's array type. The returned array will be
-            :meth:`thaw`\ ed.
+            :meth:`thaw`\ ed. When working with array containers, each leaf
+            is converted to an :class:`~numpy.ndarray`, but the container is not
+            flattened.
         """
-        pass
 
     @abstractmethod
     def to_numpy(self,
-                 array: Union[Array, ScalarLike]
-                 ) -> Union["np.ndarray[Any, Any]", ScalarLike]:
+                 array: ArrayOrContainerOrScalarT
+                 ) -> ArrayOrContainerOrScalarT:
         r"""
-        :returns: *array*, an array recognized by the context, converted
-            to a :class:`numpy.ndarray`. *array* must be
-            :meth:`thaw`\ ed.
+        :returns: an :class:`numpy.ndarray` for each array recognized by the
+            context, converted. The input *array* must be :meth:`thaw`\ ed.
+            When working with array containers, each leaf is converted to the
+            context's array type.
```

Comment:
Make it clearer that all leaves must be recognized.

Reply:
Gave it a try in https://github.com/inducer/arraycontext/compare/bb4d8fce5eec09fc59b5e125573f4ce402523ec2..8c9952aa46a6749a54330e6984d83f76c2249d7a

What do you think?